### PR TITLE
Fix hyperscan/chimera with msvc

### PR DESCRIPTION
**Pull request by syoliver**
Specify library name and version:  hyperscan/5.4.0

Hyperscan is built with a vectorized calling convention and pcre with the default calling convention on MSVC

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Where does this patch come from? Have you shared it with upstream?

**Comment by syoliver** [comment]
> Where does this patch come from?

I did it, it doesn't link because pcre is built with cdecl calling convension, but hyperscan use vector calling convention

> Have you shared it with upstream?

No, because pcre is built from source, then the calling convention is inherited from hyperscan project.

**Comment by uilianries** [comment]
@syoliver do you have error log where it fails to build without your patch?

**Comment by syoliver** [comment]
For this specific fix :

```
chimera.lib(ch_compile.obj) : error LNK2019: symbole externe non résolu pcre_compile@@40 référencé dans la fonction "public: void __cdecl `anonymous namespace'::PatternDat
a::buildPcre(char const *,unsigned int)" (?buildPcre@PatternData@?A0xa0ee20c1@@QEAAXPEBDI@Z) [C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1
950897b082206a537\build_subfolder\source_subfolder\unit\unit-chimera.vcxproj]
    Conseil sur les symboles définis et susceptibles de correspondre :
      pcre_compile
chimera.lib(ch_compile.obj) : error LNK2019: symbole externe non résolu pcre_fullinfo@@32 référencé dans la fonction "public: void __cdecl `anonymous namespace'::PatternDa
ta::buildPcre(char const *,unsigned int)" (?buildPcre@PatternData@?A0xa0ee20c1@@QEAAXPEBDI@Z) [C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd
1950897b082206a537\build_subfolder\source_subfolder\unit\unit-chimera.vcxproj]
    Conseil sur les symboles définis et susceptibles de correspondre :
      pcre_fullinfo
chimera.lib(ch_compile.obj) : error LNK2019: symbole externe non résolu pcre_study@@24 référencé dans la fonction "public: void __cdecl `anonymous namespace'::PatternData:
:buildPcre(char const *,unsigned int)" (?buildPcre@PatternData@?A0xa0ee20c1@@QEAAXPEBDI@Z) [C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd195
0897b082206a537\build_subfolder\source_subfolder\unit\unit-chimera.vcxproj]
    Conseil sur les symboles définis et susceptibles de correspondre :
      pcre_study
C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537\build_subfolder\source_subfolder\bin\unit-chimera.exe : fatal error LNK1120:
 3 externes non résolus [C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537\build_subfolder\source_subfolder\unit\unit-chimera.
vcxproj]
hyperscan/5.4.0:
hyperscan/5.4.0: ERROR: Package 'bdf07d32a526c412e696bd1950897b082206a537' build failed
hyperscan/5.4.0: WARN: Build folder C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537
ERROR: hyperscan/5.4.0: Error in build() method, line 100
        cmake.build()
        ConanException: Error 1 while executing cmake --build C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537\build_subfolder --config Debug -- /m:12 /verbosity:minimal

```

Thoses symbols exit in pcre, but with cdecl : 
- `pcre_compile` for `pcre_compile@@40`
- `pcre_fullinfo` for `pcre_fullinfo@@32`
- `pcre_study` for `pcre_study@@24`


Actually, I fixed pcre dependency too, which lead to this error before the fix :

```
LINK : fatal error LNK1104: impossible d'ouvrir le fichier 'pcre.lib' [C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537\build
_subfolder\source_subfolder\unit\unit-chimera.vcxproj]
  unit-hyperscan.vcxproj -> C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537\build_subfolder\source_subfolder\bin\unit-hypers
  can.exe
hyperscan/5.4.0:
hyperscan/5.4.0: ERROR: Package 'bdf07d32a526c412e696bd1950897b082206a537' build failed
hyperscan/5.4.0: WARN: Build folder C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537
ERROR: hyperscan/5.4.0: Error in build() method, line 100
        cmake.build()
        ConanException: Error 1 while executing cmake --build C:\Users\oliver\.conan\data\hyperscan\5.4.0\_\_\build\bdf07d32a526c412e696bd1950897b082206a537\build_subfolder --config Debug -- /m:12 /verbosity:minimal
```

But I don't think the second point is a problem.

**Comment by prince-chrismc** [comment]
Tous les problem est dans the dossier unit? https://github.com/intel/hyperscan/tree/master/unit/chimera

It's not supposed to be compiled in CCI 🤔 

It's a half surprise compiling from source plus linking cause mismatch symbols but I am surprised you needed to write a whole wrapper section

**Comment by syoliver** [comment]
> Tous les problem est dans the dossier unit? https://github.com/intel/hyperscan/tree/master/unit/chimera
> 
> It's not supposed to be compiled in CCI 🤔 

Not false, but I will have the same issue when linking my program to hyperscan. 


> It's a half surprise compiling from source plus linking cause mismatch symbols but I am surprised you needed to write a whole wrapper section 

The link mismatch is caused by the fact that hyperscan project is compiled with vectorcall option (/Gv) and that declarations in pcre header doesn't specify the calling convention (cdecl). Then msvc assume that pcre has been complied with vectorcall option too, which is not the case. 

In upstream, the sources of pcre is `add_subdirectory`ed by hyperscan. The vectorcall option which is stored in CMAKE_CXX_FLAGS is inherited and in this case pcre is built with this option too. And no mismatch occurs.

File: recipes/hyperscan/all/patches/fix-cmake.patch [hunk 1]
```diff
@@ -220,35 +220,216 @@ index 8bc6077..579a7b4 100644
...
+diff --git a/chimera/ch_pcre_bridge.h b/chimera/ch_pcre_bridge.h
+new file mode 100644
+index 0000000..bc78c7c
+--- /dev/null
++++ b/chimera/ch_pcre_bridge.h
+@@ -0,0 +1,32 @@
++#pragma once
++
++#ifndef CH_PCRE_BRIDGE_H_
++#define CH_PCRE_BRIDGE_H_
```
Comment (jwillikers) [inline recipes/hyperscan/all/patches/fix-cmake.patch:378]: An include guard and `pragma once`?

File: recipes/hyperscan/all/patches/fix-cmake.patch [hunk 2]
```diff
@@ -220,35 +220,216 @@ index 8bc6077..579a7b4 100644
...
++target_link_libraries(chimera hs PCRE::libpcre)
+ 
+ install(TARGETS chimera DESTINATION ${CMAKE_INSTALL_LIBDIR})
+ 
+@@ -44,6 +47,6 @@ if (NOT WIN32)
      set(PRIVATE_LIBS "${PRIVATE_LIBS} -L${LIBDIR} -lpcre")
  
      configure_file(libch.pc.in libch.pc @ONLY) # only replace @ quoted vars
 -    install(FILES ${CMAKE_BINARY_DIR}/chimera/libch.pc
 +    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/libch.pc
```
Comment (jwillikers) [inline recipes/hyperscan/all/patches/fix-cmake.patch:256]: There's no reason to fix this install command since the file is deleted in the `conanfile.py` package step anyways. Can you just comment out the `install` command here?

**Review by jwillikers — CHANGES_REQUESTED** [review]

File: recipes/hyperscan/all/patches/fix-cmake.patch (hunk 1 above)
Comment (syoliver) [reply recipes/hyperscan/all/patches/fix-cmake.patch:378]: Yes, it is usual way to do: ```pragma once``` before include guards for compatibility.

File: recipes/hyperscan/all/patches/fix-cmake.patch (hunk 2 above)
Comment (syoliver) [reply recipes/hyperscan/all/patches/fix-cmake.patch:256]: True, I will fix it as soon as possible. Thanks !

**Comment by syoliver** [comment]
> It's not supposed to be compiled in CCI 🤔

@prince-chrismc Removed unit tests.

File: recipes/hyperscan/all/patches/fix-cmake.patch (hunk 2 above)
Comment (syoliver) [reply recipes/hyperscan/all/patches/fix-cmake.patch]: Fixed

**Review by jwillikers — DISMISSED** [review]

**Comment by prince-chrismc** [comment]
Only think missing is https://github.com/conan-io/conan-center-index/blob/bcb13631c96d8e93741fdc5e057d9c8ad62248e2/recipes/hyperscan/all/conandata.yml needs  https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

This is a pretty complicated patch for vendoring a dep and it's worth writing down i think :)

File: recipes/hyperscan/all/conanfile.py [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
+from conan.tools.files import apply_conandata_patches, rename, get, copy, rmdir
+from conan.tools.scm import Version
+from conan.tools.build import check_min_cppstd
 
 import os
 
-required_conan_version = ">=1.43.0"
+required_conan_version = ">=1.53.0"
```
Comment (uilianries) [inline recipes/hyperscan/all/conanfile.py:10]: Any motivation to require 1.53.0 here?

Comment (syoliver) [reply recipes/hyperscan/all/conanfile.py:10]: I don't know what version is needed by the rewrite I did for conan V2. I just tested with 1.53.0.
You have an advice ?

File: recipes/hyperscan/all/conanfile.py [hunk 4]
```diff
@@ -41,27 +45,40 @@ class HyperscanConan(ConanFile):
...
 
     def requirements(self):
-        self.requires("boost/1.79.0");
+        self.requires("boost/1.79.0")
         if self.options.build_chimera:
             self.requires("pcre/8.45")
 
     def validate(self):
-        tools.check_min_cppstd(self, "11")
+        if self.settings.compiler.get_safe("cppstd"):
```
Comment (uilianries) [inline recipes/hyperscan/all/conanfile.py:74]:
```suggestion
        if self.info.settings.compiler.get_safe("cppstd"):
```
It's part of the migration guide: https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#the-validate-method

File: recipes/hyperscan/all/conanfile.py [hunk 5]
```diff
@@ -1,9 +1,13 @@
...
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.cmake import CMake, CMakeToolchain
+from conan.tools.files import apply_conandata_patches, get, copy, rmdir
+from conan.tools.scm import Version
+from conan.tools.build import check_min_cppstd
 
 import os
 
-required_conan_version = ">=1.43.0"
+required_conan_version = ">=1.53.0"
```
Comment (uilianries) [inline recipes/hyperscan/all/conanfile.py:10]:
```suggestion
required_conan_version = ">=1.51.1"
```

Because you need self.info.settings.get_safe: https://docs.conan.io/en/latest/changelog.html#id63

File: recipes/hyperscan/all/conanfile.py [hunk 6]
```diff
@@ -41,27 +45,40 @@ class HyperscanConan(ConanFile):
...
+            check_min_cppstd(self, "11")
+
+        minimum_version = self._minimum_compilers_version.get(str(self.info.settings.compiler), False)
+        if not minimum_version:
+            self.output.warn("{}/{} requires C++11. Your compiler is unknown. Assuming it supports C++11.".format(self.name, self.version))
+        elif Version(self.info.settings.compiler.version) < minimum_version:
+            raise ConanInvalidConfiguration("{}/{} requires C++11, which your compiler does not support.".format(self.name, self.version))
 
         if self.settings.arch not in ["x86", "x86_64"]:
             raise ConanInvalidConfiguration("Hyperscan only support x86 architecture")
```
Comment (prince-chrismc) [inline recipes/hyperscan/all/conanfile.py:84]: message is a little confusing

File: recipes/hyperscan/all/conanfile.py [hunk 7]
```diff
@@ -94,17 +108,17 @@ def configure(self):
             raise ConanInvalidConfiguration("Chimera build requires static building")
```
Comment (prince-chrismc) [inline recipes/hyperscan/all/conanfile.py:108]: This should be in validate for the best results

File: recipes/hyperscan/all/patches/fix-cmake.patch [hunk 8]
```diff
@@ -67,10 +67,11 @@ index 8bc6077..579a7b4 100644
...
  endif()
-@@ -483,16 +483,16 @@ if (CORRECT_PCRE_VERSION AND PCRE_BUILD_SOURCE AND BUILD_STATIC_LIBS)
+@@ -482,33 +482,18 @@ if (CORRECT_PCRE_VERSION AND PCRE_BUILD_SOURCE AND BUILD_STATIC_LIBS)
+     set(BUILD_CHIMERA TRUE)
  endif()
  
- add_subdirectory(unit)
+-add_subdirectory(unit)
 -if (EXISTS ${CMAKE_SOURCE_DIR}/tools/CMakeLists.txt)
 +if (EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/tools/CMakeLists.txt)
```
Comment (prince-chrismc) [inline recipes/hyperscan/all/patches/fix-cmake.patch:76]: Diff in diff view is hard to read --- but I think the old one is not require since the new toolchain does not add subdirectory

minor comment (this new code is better) just trying to cleanup

File: recipes/hyperscan/all/patches/fix-cmake.patch [hunk 9]
```diff
@@ -220,35 +229,226 @@ index 8bc6077..579a7b4 100644
...
+--- a/chimera/ch_database.h
++++ b/chimera/ch_database.h
+@@ -38,9 +38,7 @@ extern "C"
  {
  #endif
  
 -#define PCRE_STATIC
+-#include <pcre.h>
+-
++#include "ch_pcre_bridge.h"
```
Comment (prince-chrismc) [inline recipes/hyperscan/all/patches/fix-cmake.patch:337]: This should be a separate patch for visibility

**Review by prince-chrismc — COMMENTED** [review]
Handful of v2 comments you might want to check the template https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package there a few things missing

one thing not in my comments int the `test_v1_package` is also needed
